pyccl/p2d.py

```diff
@@ -0,0 +1,142 @@
+from . import ccllib as lib
+
+from .core import check
+import numpy as np
+
+
+class Pk2D(object):
+    """A power spectrum class holding the information needed to reconstruct an
+    arbitrary function of wavenumber and scale factor.
+
+    Args:
+        pkfunc (:obj:function): a function returning a floating point
+             number or numpy array with the signature `f(k,a)`, where k
+             is a wavenumber (in units of Mpc^-1) and a is the scale
+             factor. The function must able to take numpy arrays as `k`.
+             The function must return the value(s) of the power spectrum
+             (or its natural logarithm, depending on the value of
+             `is_logp`. The power spectrum units should be compatible
+             with those used by CCL (e.g. if you're passing a matter power
+             spectrum, its units should be Mpc^3). If this argument is not
+             `None`, this function will be sampled at the values of k and
+             a used internally by CCL to store the linear and non-linear
+             power spectra.
+        a_arr (array): an array holding values of the scale factor
+        lk_arr (array): an array holding values of the natural logarithm
+             of the wavenumber (in units of Mpc^-1).
+        pk_arr (array): a 2D array containing the values of the power
+             spectrum at the values of the scale factor and the wavenumber
+             held by `a_arr` and `lk_arr`. The shape of this array must be
+             `[na,nk]`, where `na` is the size of `a_arr` and `nk` is the
+             size of `lk_arr`. This array can be provided in a flattened
+             form as long as the total size matches `nk*na`. The array can
+             hold the values of the natural logarithm of the power
+             spectrum, depending on the value of `is_logp`. If `pkfunc`
+             is not None, then `a_arr`, `lk_arr` and `pk_arr` are ignored.
+             However, either `pkfunc` or all of the last three array must
+             be non-None. Note that, if you pass your own Pk array, you
+             are responsible of making sure that it is sufficiently well
+             sampled (i.e. the resolution of `a_arr` and `lk_arr` is high
+             enough to sample the main features in the power spectrum).
+             For reference, CCL will use bicubic interpolation to evaluate
+             the power spectrum at any intermediate point in k and a.
+        interp_order_lok (int): extrapolation order to be used on k-values
+             below the minimum of the splines (use 0, 1 or 2).
+        interp_order_hik (int): extrapolation order to be used on k-values
+             above the maximum of the splines (use 0, 1 or 2).
```

Comment:
extrapolation is done fitting a polynomial. Is this in P-k-a space, or logP-logk-loga space? Or is this determined by 'islogP'?

Reply:
It's determined by `islogP` precisely to avoid that problem. I'll document better.

Reply:
So if I want log-log interpolation to be done, I have to pass in all log values at the start? This is fine, but should be well documented. (It's the sort of thing I have messed up in the past.)

Reply:
Is that not clear enough from the description of `is_logp` ?

Reply:
What is expected is indeed clear. The fact that log interpolation can only be done this way is less clear.

Reply:
OK, just clarified that.